// !WITH_NEW_INFERENCE

annotation class A1
annotation class A2(val some: Int = 12)

val <<!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A1<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A2(3)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A1(<!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>12<!><!>)<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2(<!NI;TYPE_MISMATCH!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>"Test"<!><!><!>)<!><!><!><!>  T> T.topProp: Int get() = 12

class SomeClass {
    val <<!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A1<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@A2(3)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A1(<!NI;TOO_MANY_ARGUMENTS!><!TOO_MANY_ARGUMENTS!>12<!><!>)<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!NI;REPEATED_ANNOTATION!><!WRONG_ANNOTATION_TARGET!><!REPEATED_ANNOTATION!>@A2(<!NI;TYPE_MISMATCH!><!NI;TYPE_MISMATCH!><!TYPE_MISMATCH!>"Test"<!><!><!>)<!><!><!><!> T> T.field: Int get() = 12

    fun foo() {
        val <<!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>A1<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>A2<!><!>(3)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>A2<!><!><!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>A1<!><!>(12)<!><!> <!NI;WRONG_ANNOTATION_TARGET!><!WRONG_ANNOTATION_TARGET!>@<!NI;DEBUG_INFO_MISSING_UNRESOLVED!><!DEBUG_INFO_MISSING_UNRESOLVED!>A2<!><!>("Test")<!><!> T> <!NI;UNUSED_VARIABLE!><!UNUSED_VARIABLE!>localVal<!><!> = 12
    }
}